llm-finetuning/huggingface/hf_model_deployer.py

```diff
@@ -0,0 +1,364 @@
+from zenml.services import BaseService, ServiceConfig
+from huggingface.hf_deployment_service import (
+    HuggingFaceDeploymentService,
+    HuggingFaceServiceConfig,
+)
+from huggingface.hf_model_deployer_flavor import (
+    HuggingFaceModelDeployerSettings,
+    HuggingFaceModelDeployerConfig,
+)
+from zenml.model_deployers.base_model_deployer import (
+    DEFAULT_DEPLOYMENT_START_STOP_TIMEOUT,
+    BaseModelDeployerFlavor,
+)
+from zenml.client import Client
+from zenml.services import ServiceRegistry
+from zenml.artifacts.utils import save_artifact, log_artifact_metadata
+
+logger = get_logger(__name__)
+
+
+class HuggingFaceModelDeployer(BaseModelDeployer):
+    """Huggingface endpoint model deployer."""
+
+    NAME: ClassVar[str] = "HFEndpoint"
+    FLAVOR: ClassVar[
+        Type[BaseModelDeployerFlavor]
+    ] = HuggingFaceModelDeployerFlavor
+
+    @property
+    def config(self) -> HuggingFaceModelDeployerConfig:
+        """Config class for the Huggingface Model deployer settings class.
+
+        Returns:
+            The configuration.
+        """
+        return cast(HuggingFaceModelDeployerConfig, self._config)
+
+    @property
+    def settings_class(self) -> Type[HuggingFaceModelDeployerSettings]:
+        """Settings class for the Huggingface Model deployer settings class.
+
+        Returns:
+            The settings class.
+        """
+        return HuggingFaceModelDeployerSettings
+
+    def _create_new_service(
+        self, timeout: int, config: HuggingFaceServiceConfig
+    ) -> HuggingFaceDeploymentService:
+        """Creates a new HuggingFaceDeploymentService.
+
+        Args:
+            timeout: the timeout in seconds to wait for the Huggingface inference endpoint
+                to be provisioned and successfully started or updated.
+            config: the configuration of the model to be deployed with Huggingface model deployer.
+
+        Returns:
+            The HuggingFaceServiceConfig object that can be used to interact
+            with the Huggingface inference endpoint.
+        """
+        # create a new service for the new model
+        service = HuggingFaceDeploymentService(config)
+
+        service_metadata = service.dict()
+        # Use first 8 characters of UUID as artifact version
+        artifact_version = str(service_metadata["uuid"])[:8]
+
+        # Add first 8 characters of UUID to endpoint name
+        service.config.endpoint_name += artifact_version
+
+        service_metadata = service.dict()
+        save_artifact(
+            service,
+            "hf_deployment_service",
+            version=artifact_version,
+            is_deployment_artifact=True,
+        )
+        # UUID object is not json serializable
+        service_metadata["uuid"] = str(service_metadata["uuid"])
+        log_artifact_metadata(
+            artifact_name="hf_deployment_service",
+            artifact_version=artifact_version,
+            metadata={"hf_deployment_service": service_metadata},
```

Comment:
maybe we can store `service_metadata.dict()` (or even only the data we need) instead of this object? Might make things cleaner in the below functions where we fetch it